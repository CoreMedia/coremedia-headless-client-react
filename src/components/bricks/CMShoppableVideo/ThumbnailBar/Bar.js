// @flow
import styled from 'styled-components';
import theme from 'styled-theming';

const color = theme('name', {
  default: 'rgba(0, 108, 174, 1)',
  hkm: '#fabfc0',
});

const Bar = styled.hr.attrs({
  'aria-hidden': true,
})`
  position: relative;
  left: ${props => `${props.posX}px`};
  width: ${props => `${props.width}px`};
  margin: 0;
  border-width: 0 0 3px;
  border-style: none none solid;
  border-color: ${props => (props.width ? color : 'transparent')};
  transition: ${props => (props.posX && props.width ? 'all' : 'border-color')} 0.6s ease-out 0s;
`;
Bar.displayName = 'Bar';

/*Bar.propTypes = {
  posX: PropTypes.number.isRequired,
  width: PropTypes.number.isRequired,
  theme: PropTypes.shape({
    name: PropTypes.string.isRequired,
  }),
};*/

Bar.defaultProps = {
  theme: {
    name: 'default',
  },
};

export default Bar;
