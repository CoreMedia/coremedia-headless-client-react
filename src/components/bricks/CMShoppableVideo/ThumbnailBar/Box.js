// @flow
import styled from 'styled-components';

const ThumbnailsBox = styled.div`
  display: inline-block;
  white-space: nowrap;
`;

ThumbnailsBox.displayName = 'ThumbnailsBox';

export default ThumbnailsBox;
