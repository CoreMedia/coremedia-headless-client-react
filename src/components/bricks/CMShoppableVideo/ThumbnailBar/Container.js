// @flow
import type { ComponentType } from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';

type Props = {
  translateX: number,
};

const ThumbnailsContainer: ComponentType<Props> = styled.div`
  transform: translateX(${props => `${props.translateX}px`});
  transition: transform 0.45s ease-out;
`;

ThumbnailsContainer.displayName = 'ThumbnailsContainer';

ThumbnailsContainer.propTypes = {
  translateX: PropTypes.number.isRequired,
};

export default ThumbnailsContainer;
