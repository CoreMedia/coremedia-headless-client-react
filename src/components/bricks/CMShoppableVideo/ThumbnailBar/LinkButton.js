// @flow
import type { ComponentType } from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';

type Props = {
  active: boolean,
  ariaLabel?: string,
};

const LinkButton: ComponentType<Props> = styled.a.attrs({
  role: 'button',
  'aria-label': props => props.ariaLabel,
})`
  display: inline-block;
  opacity: ${props => (props.active ? 1 : 0.4)};
  transition: opacity 0.3s linear 0.3s;
`;
LinkButton.displayName = 'LinkButton';

LinkButton.propTypes = {
  active: PropTypes.bool.isRequired,
  ariaLabel: PropTypes.string,
};

export default LinkButton;
