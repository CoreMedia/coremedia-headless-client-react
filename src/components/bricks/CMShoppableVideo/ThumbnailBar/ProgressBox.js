// @flow
import styled from 'styled-components';

const ProgressBox = styled.div`
  width: 100%;
`;
ProgressBox.displayName = 'ProgressBox';

export default ProgressBox;
