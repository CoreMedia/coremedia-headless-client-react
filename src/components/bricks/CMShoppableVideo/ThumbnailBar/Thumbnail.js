// @flow
import React from 'react';
import PropTypes from 'prop-types';

import LinkButton from './LinkButton';
import Picture from '../../CMPicture';

type Props = {
  ariaLabel?: string,
  pictureLink: string,
  pictureTitle: string,
  pictureAlt: string,
  handleClick: (ev: SyntheticEvent<HTMLButtonElement>) => void,
  active: boolean,
};

class Thumbnail extends React.Component<Props> {
  static propTypes = {
    ariaLabel: PropTypes.string,
    pictureLink: PropTypes.string.isRequired,
    pictureTitle: PropTypes.string.isRequired,
    pictureAlt: PropTypes.string.isRequired,
    handleClick: PropTypes.func.isRequired,
    active: PropTypes.bool.isRequired,
  };

  shouldComponentUpdate(nextProps: Props) {
    return (
      nextProps.ariaLabel !== this.props.ariaLabel ||
      nextProps.pictureLink !== this.props.pictureLink ||
      nextProps.pictureTitle !== this.props.pictureTitle ||
      nextProps.pictureAlt !== this.props.pictureAlt ||
      nextProps.active !== this.props.active
    );
  }

  render() {
    const { ariaLabel, pictureLink, pictureTitle, pictureAlt, handleClick, active } = this.props;
    return (
      <LinkButton active={active} ariaLabel={ariaLabel} onClick={handleClick}>
        <Picture link={pictureLink} ratio="thumbnail" title={pictureTitle} alt={pictureAlt} />
      </LinkButton>
    );
  }
}

export default Thumbnail;
