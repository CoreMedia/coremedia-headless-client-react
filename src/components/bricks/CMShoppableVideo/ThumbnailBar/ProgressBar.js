// @flow
import React from 'react';
import PropTypes from 'prop-types';

import ProgressBox from './ProgressBox';
import Bar from './Bar';

type Props = {
  posX: number,
  width: number,
};

const ProgressBar = ({ posX, width }: Props) => (
  <ProgressBox>
    <Bar posX={posX} width={width} />
  </ProgressBox>
);

ProgressBar.propTypes = {
  posX: PropTypes.number,
  width: PropTypes.number,
};

ProgressBar.defaultProps = {
  posX: 0,
  width: 0,
};

export default ProgressBar;
