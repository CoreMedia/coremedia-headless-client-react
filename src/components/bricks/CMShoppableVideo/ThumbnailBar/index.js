// @flow
import React from 'react';
import PropTypes from 'prop-types';

import Thumbnail from './Thumbnail';
import ProgressBar from './ProgressBar';
import Container from './Container';
import Wrapper from './Wrapper';
import Box from './Box';

type Props = {
  thumbnails: Array<Object>,
  translateX: number,
  handleRef: (element: ?HTMLElement) => HTMLElement | null | void,
  prevItems: number,
  range: number,
  thumbWidth: number,
};

const ThumbnailBar = ({
  thumbnails,
  translateX,
  handleRef,
  prevItems,
  range,
  thumbWidth,
}: Props) => {
  const posX = prevItems * thumbWidth;
  const width = range * thumbWidth;

  return (
    <Wrapper>
      <Container translateX={translateX}>
        <Box innerRef={handleRef}>
          {thumbnails.map(item => (
            <Thumbnail
              key={item.id}
              active={item.active}
              ariaLabel={`Open product preview for "${item.teaserTitle}"`}
              handleClick={item.handleClick}
              pictureLink={item.pictureLink}
              pictureTitle={item.pictureTitle}
              pictureAlt={item.pictureAlt}
            />
          ))}
        </Box>
        <ProgressBar posX={posX} width={width} />
      </Container>
    </Wrapper>
  );
};

ThumbnailBar.propTypes = {
  thumbnails: PropTypes.arrayOf(PropTypes.object).isRequired,
  translateX: PropTypes.number.isRequired,
  handleRef: PropTypes.func.isRequired,
  prevItems: PropTypes.number.isRequired,
  range: PropTypes.number.isRequired,
  thumbWidth: PropTypes.number.isRequired,
};

export default ThumbnailBar;
