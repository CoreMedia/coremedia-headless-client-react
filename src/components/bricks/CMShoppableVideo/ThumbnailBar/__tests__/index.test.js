// @flow
import React from 'react';

import ThumbnailBar from '..';

describe('ThumbnailBar Component', () => {
  it('should render correctly', () => {
    const handleClick = jest.fn().mockName('handleClick');
    const handleRef = jest.fn().mockName('handleRef');
    const wrapper = shallow(
      <ThumbnailBar
        thumbnails={[
          {
            id: 'id123',
            active: true,
            teaserTitle: 'Title',
            pictureLink: 'coremedia:///image/6188/data',
            pictureTitle: 'picture title',
            pictureAlt: 'alt',
            handleClick,
          },
        ]}
        translateX={0}
        handleRef={handleRef}
        prevItems={0}
        range={0}
        thumbWidth={78}
      />
    );
    expect(wrapper).toMatchSnapshot();
  });
  it('should calculate posX and width for ProgressBar correctly', () => {
    const handleClick = jest.fn().mockName('handleClick');
    const handleRef = jest.fn().mockName('handleRef');
    const wrapper = shallow(
      <ThumbnailBar
        thumbnails={[
          {
            id: 'id123',
            active: true,
            teaserTitle: 'Title',
            pictureLink: 'coremedia:///image/6188/data',
            pictureTitle: 'picture title',
            pictureAlt: 'alt',
            handleClick,
          },
        ]}
        translateX={0}
        handleRef={handleRef}
        prevItems={1}
        range={2}
        thumbWidth={78}
      />
    );
    expect(wrapper).toMatchSnapshot();
  });
});
