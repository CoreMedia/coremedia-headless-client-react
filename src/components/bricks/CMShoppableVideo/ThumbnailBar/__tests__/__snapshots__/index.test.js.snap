// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ThumbnailBar Component should calculate posX and width for ProgressBar correctly 1`] = `
<ThumbnailsWrapper>
  <ThumbnailsContainer
    translateX={0}
  >
    <ThumbnailsBox
      innerRef={[MockFunction handleRef]}
    >
      <Thumbnail
        active={true}
        ariaLabel="Open product preview for \\"Title\\""
        handleClick={[MockFunction handleClick]}
        key="id123"
        pictureAlt="alt"
        pictureLink="coremedia:///image/6188/data"
        pictureTitle="picture title"
      />
    </ThumbnailsBox>
    <ProgressBar
      posX={78}
      width={156}
    />
  </ThumbnailsContainer>
</ThumbnailsWrapper>
`;

exports[`ThumbnailBar Component should render correctly 1`] = `
<ThumbnailsWrapper>
  <ThumbnailsContainer
    translateX={0}
  >
    <ThumbnailsBox
      innerRef={[MockFunction handleRef]}
    >
      <Thumbnail
        active={true}
        ariaLabel="Open product preview for \\"Title\\""
        handleClick={[MockFunction handleClick]}
        key="id123"
        pictureAlt="alt"
        pictureLink="coremedia:///image/6188/data"
        pictureTitle="picture title"
      />
    </ThumbnailsBox>
    <ProgressBar
      posX={0}
      width={0}
    />
  </ThumbnailsContainer>
</ThumbnailsWrapper>
`;
